codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java

```diff
@@ -223,70 +264,240 @@ private int pathEndIdx() {
+            if (indexOfEquals != -1) {
+                // we have found `=` first (which is quite common); we can drop on check
+                nextValueStart = indexOfEquals + 1;
+                // no need for a switch; we bias our decisions based on the current state
+                for (int i = nextValueStart; i < len; i++) {
+                    char ch = s.charAt(i);
+                    if (ch == '&' || (!semicolonIsNormalChar && ch == ';')) {
+                        valueEndExclusive = i;
+                        break;
+                    }
+                    if (ch == '#') {
+                        len = i;
+                        break;
                     }
                 }
-                nameStart = i + 1;
+            }
+            if (valueEndExclusive == -1) {
+                valueEndExclusive = len;
+            }
+            int valueStart;
+            if (nextValueStart != -1) {
+                valueStart = nextValueStart;
+                if (valueStart == nameStart + 1) {
+                    // uncommon slow path: it seems there is no name!
+                    // search nameStart while skipping useless subsequent =, if any
+                    nameStart = skipIf(s, valueStart, valueEndExclusive, '=');
+                    valueStart = indexOf(s, nameStart + 1, valueEndExclusive, '=');
+                }
+            } else {
+                valueStart = -1;
+            }
+            addParam(s, nameStart, valueStart, valueEndExclusive, collector, accumulator, charset);
+            if (valueEndExclusive == len) {
                 break;
-            case '#':
-                break loop;
-            default:
-                // continue
             }
+            nameStart = valueEndExclusive + 1;
         }
-        addParam(s, nameStart, valueStart, i, params, charset);
-        return params;
     }
 
-    private static boolean addParam(String s, int nameStart, int valueStart, int valueEnd,
-                                    Map<String, List<String>> params, Charset charset) {
+    private static int indexOf(String s, int from, int to, int ch) {
+        for (int i = from; i < to; i++) {
+            if (s.charAt(i) == ch) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private static int skipIf(String s, int from, int to, int ch) {
+        for (int i = from; i < to; i++) {
+            if (s.charAt(i) != ch) {
+                return i;
+            }
+        }
+        return to;
+    }
+
+    private static final ParameterCollector<Map<String, List<String>>> MAP_PARAMETER_COLLECTOR =
+            new ParameterCollector<Map<String, List<String>>>() {
+                @Override
+                public void accept(String name, String value, Map<String, List<String>> accumulator) {
+                    List<String> values = accumulator.get(name);
+                    if (values == null) {
+                        values = new ArrayList<String>(1);
```

Comment:
is 1 a good default ? Maybe use 4 so we dont need to expand directly

Reply:
I've no idea - I would expect that query params usually have a single value for each param, most of the time, rather then more, but I have not much exp in it